plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/ForceApply.hs

```diff
@@ -0,0 +1,186 @@
+{- | The 'ForceApply' optimisation pushes 'Force' inside its direct 'Apply' subterms,
+ removing any 'Delay' at the top of the body of the underlying lambda abstraction.
+ For example, @force ((\x -> delay b) a)@ is transformed into @(\x -> b) a@.
+ In such simple cases, the transformation is obviously correct, the question remains
+ if this approach can be generalised.
+
+ Since UPLC programs are created from erasing the types of TPLC programs (see
+ "PlutusCore.Compiler.Erase") we will consider TPLC terms of the following structure:
+
+ @
+   /\T1 -> \X1 -> /\T2 -> \X2 -> /\T3 -> \X3 -> ... -> /\Tn -> \Xn -> body
+ @
+ where @T1 ... Tn@ are lists of type variables (e.g. @T1@ could be @[t, q, p]@)
+ and @X1 ... Xn@ are lists of term variables. Of course, each @/\@ and @\@ here would
+ desugar to a sequence of type/term abstractions.
+
+ In order to reason about the proposed optimisation we need to consider such terms in
+ the context of them being applied to some sequence of terms.
+
+ One important observation is that this transformation requires that the underlying
+ (term)-lambda abstraction will be fully reduced by the applications.
+ For UPLC, this can happen only when the number of lambda abstracted variables is equal
+ to the number of terms to which it will be applied.
+ For example, @force (\x -> delay b) => (\x -> b)@ is invalid, since the former is @error@.
+ The other case, we can see that applying the optimisation modifies the end result:
+ @
+   force ((\x -> delay b) a1 a2) => (\x -> b) a1 a2 => b[x1/a1] a2
+
+   vs.
+
+   force ((\x -> delay b) a1 a2) => force ((delay b[x1/a1]) a2)
+ @
+ The latter cannot be reduced further according to the reducation semantics of UPLC.
+
+ To generalise, we consider the family of terms above applied to a family of types and
+ terms:
+ @
+   (/\T1 -> \X1 ->  ... -> /\Tn -> \Xn -> body)
+     T1 X1 ... Tn Xn
+ @
+ For brevity, the types and the terms to which the lambda applies are named the same as the
+ bound variables, but of course this isn't necessary. Also note that in general @|Tn| == |Xn|@
+ doesn't necessarily hold for any @n@.
+
+ Translated to UPLC, the original term is:
+ @
+   delay^|T1| (\X1 -> delay^|T2| (\X2 -> delay^|T3| (\X3 -> ... -> delay^|Tn| (\Xn -> body))))
+ @
+ where @delay^|A|@ means "apply delay |A| (the length of A) times".
+ With the applications:
+ @
+   (force^|Tn| (... (force^|T3| ((force^|T2| ((force^|T1| original) X1)) X2)) X3 ...)) Xn
+ @
+ After inlining @original@ we get:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((force^|T2|
+           ((force^|T1|
```

Comment:
Why do we jump from one opening paren to two opening parens and then stay at two opening parens? I tried writing down the expression myself and only got one pair of parens at each node, but I'm known for messing up such things.

Reply:
Sorry, I think I messed this up. Will fix, thanks.

Reply:
Actually I don't think I messed this up. The reason there are two parenthesis before each `force` is that we have `(force (something something))` applied to `Xi` so `((force (something something)) Xi)`. 
I omitted one of the parenthesis after, respectively before, the two `...`.

Reply:
So, there should be two parenthesis before each `force`. Notice that in my fix `X3` is followed by a single parenthesis because we're not explicitly showing the application to `X4`. I hope this makes sense, it's confusing for me as well so I'm not ruling out being wrong about this as well.